radixdlt-engine/src/main/java/com/radixdlt/application/system/scrypt/SystemConstraintScrypt.java

```diff
@@ -0,0 +1,169 @@
+		public REAddrClaimStart(byte[] arg) {
+			this.arg = arg;
+		}
+
+		public ReducerState claim(UnclaimedREAddr unclaimedREAddr, ExecutionContext ctx) throws ProcedureException {
+			if (ctx.permissionLevel() != PermissionLevel.SYSTEM
+				&& !ctx.key().map(k -> unclaimedREAddr.allow(k, arg)).orElse(false)) {
+				throw new ProcedureException("Invalid key/arg combination.");
+			}
+			return new REAddrClaim(unclaimedREAddr, arg);
+		}
+	}
+
+	@Override
+	public void main(Loader os) {
+		os.procedure(
+			new SystemCallProcedure<>(
+				TokenHoldingBucket.class, REAddr.ofSystem(),
+				() -> new Authorization(PermissionLevel.USER, (r, c) -> { }),
+				(s, d, c) -> {
+					var id = d.get(0);
+					var syscall = Syscall.of(id).orElseThrow(() -> new ProcedureException("Invalid call type " + id));
+					if (syscall != Syscall.FEE_RESERVE_PUT) {
+						throw new ProcedureException("Invalid call type: " + syscall);
+					}
+
+					var amt = d.getUInt256(1);
+					var tokens = s.withdraw(REAddr.ofNativeToken(), amt);
+					c.depositFeeReserve(tokens);
+					return ReducerResult.incomplete(s);
+				}
+			)
+		);
+
+		os.procedure(
+			new SystemCallProcedure<>(
+				VoidReducerState.class, REAddr.ofSystem(),
+				() -> new Authorization(PermissionLevel.USER, (r, c) -> { }),
+				(s, d, c) -> {
+					var id = d.get(0);
+					var syscall = Syscall.of(id).orElseThrow(() -> new ProcedureException("Invalid call type " + id));
+					if (syscall == Syscall.FEE_RESERVE_TAKE) {
+						var amt = d.getUInt256(1);
+						var tokens = c.withdrawFeeReserve(amt);
+						return ReducerResult.incomplete(new TokenHoldingBucket(tokens));
+					} else if (syscall == Syscall.READDR_CLAIM) {
+						var bytes = d.getBytes(1);
+						var str = new String(bytes);
+						if (systemNames.contains(str) && c.permissionLevel() != PermissionLevel.SYSTEM) {
+							throw new ProcedureException("Not allowed to use name " + str);
+						}
+						if (!NAME_PATTERN.matcher(str).matches()) {
+							throw new ProcedureException("invalid rri name: " + str);
+						}
+
+						return ReducerResult.incomplete(new REAddrClaimStart(bytes));
+					} else {
+						throw new ProcedureException("Invalid call type: " + syscall);
+					}
+				}
+			)
+		);
+
+		// PUB_KEY type is already claimed by accounts
+		var claimableAddrTypes =
+			EnumSet.of(REAddr.REAddrType.NATIVE_TOKEN, REAddr.REAddrType.HASHED_KEY, REAddr.REAddrType.SYSTEM);
+		os.substate(
+			new SubstateDefinition<>(
+				UnclaimedREAddr.class,
+				SubstateTypeId.UNCLAIMED_READDR.id(),
+				buf -> {
+					throw new UnsupportedOperationException();
+				},
+				(s, buf) -> {
+					throw new UnsupportedOperationException();
+				},
+				buf -> {
+					var addr = REFieldSerialization.deserializeREAddr(buf, claimableAddrTypes);
+					return new UnclaimedREAddr(addr);
+				},
+				(a, buf) -> REFieldSerialization.serializeREAddr(buf, (REAddr) a)
+			));
+
+		os.procedure(new DownProcedure<>(
+			REAddrClaimStart.class, UnclaimedREAddr.class,
```

Comment:
Could you help me understand how we ensure each REAddress is claimed once only?

Reply:
Each REAddr has one and only one virtual `UnclaimedREAddr` substate. When claimed that substate is shutdown and thus can never be claimed again.